File: server/apps/nexotype/schemas/lims/assay_run_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class AssayRunCreate(BaseModel):
    """Schema for creating a new assay run"""
    protocol_id: int
    run_date: datetime
    operator_id: int | None = None


class AssayRunUpdate(BaseModel):
    """Schema for updating an assay run"""
    protocol_id: int = None
    run_date: datetime = None
    operator_id: int | None = None


# ==========================================
# Response Schemas
# ==========================================

class AssayRunDetail(BaseModel):
    """Schema for assay run details"""
    id: int
    protocol_id: int
    run_date: datetime
    operator_id: int | None = None
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class AssayRunListResponse(BaseModel):
    """Response schema for listing assay runs"""
    success: bool
    data: list[AssayRunDetail] | None = None
    count: int = 0
    error: str | None = None


class AssayRunResponse(BaseModel):
    """Response schema for single assay run operations"""
    success: bool
    data: AssayRunDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/lims/assay-run.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const AssayRunSchema = z.object({
  id: z.number(),
  protocol_id: z.number().int(),
  run_date: z.string(),
  operator_id: z.number().int().nullable().optional(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateAssayRunSchema = z.object({
  protocol_id: z.number().int(),
  run_date: z.string().min(1),
  operator_id: z.number().int().nullable().optional(),
});

export const UpdateAssayRunSchema = z.object({
  protocol_id: z.number().int().optional(),
  run_date: z.string().min(1).optional(),
  operator_id: z.number().int().nullable().optional(),
});

// ==========================================
// Types
// ==========================================

export type AssayRun = z.infer<typeof AssayRunSchema>;
export type CreateAssayRun = z.infer<typeof CreateAssayRunSchema>;
export type UpdateAssayRun = z.infer<typeof UpdateAssayRunSchema>;

// ==========================================
// Response Types
// ==========================================

export type AssayRunResponse = {
  success: boolean;
  data?: AssayRun;
  error?: string;
};

export type AssayRunsResponse = {
  success: boolean;
  data?: AssayRun[];
  count?: number;
  error?: string;
};
